sdks/go/pkg/beam/core/runtime/exec/plan.go

```diff
@@ -131,6 +141,41 @@ func (p *Plan) Execute(ctx context.Context, id string, manager DataContext) erro
 	return nil
 }
 
+func (p *Plan) Finalize(ctx context.Context, id string) error {
+	if p.status != Up {
+		return errors.Errorf("invalid status for plan %v: %v", p.id, p.status)
+	}
+	failedIndices := []int{}
+	for idx, bfc := range p.bf.callbacks {
+		if time.Now().Before(bfc.validUntil) {
+			if err := bfc.callback(); err != nil {
+				failedIndices = append(failedIndices, idx)
+			}
+		}
+	}
+
+	newFinalizer := bundleFinalizer{
```

Comment:
Not sure if I missed something, but it looks like newFinalizer doesn't actually do anything. It gets populated with the failed callbacks, but then this function just returns without actually using it or storing it anywhere.

Reply:
Ooh, good catch - this should be getting assigned to `p.bf` so that if the runner tries to rerun finalization the failed callbacks get rerun, but it got dropped when I moved this code. I added it back.